File: tui/src/scala/tui/widgets/ChartWidget.scala
package tui
package widgets

import tui.Style
import tui.internal.ranges
import tui.internal.saturating._
import tui.widgets.canvas.{CanvasWidget, Line, Points}

/** A widget to plot one or more dataset in a cartesian coordinate system
  *
  * @param datasets
  *   A reference to the datasets
  * @param block
  *   A block to display around the widget eventually
  * @param xAxis
  *   The horizontal axis
  * @param yAxis
  *   The vertical axis
  * @param style
  *   The widget base style
  * @param hiddenLegendConstraints
  *   Set the constraints used to determine whether the legend should be shown or not.
  */
case class ChartWidget(
    datasets: Array[ChartWidget.Dataset],
    xAxis: ChartWidget.Axis = ChartWidget.Axis.default,
    yAxis: ChartWidget.Axis = ChartWidget.Axis.default,
    style: Style = Style.DEFAULT,
    hiddenLegendConstraints: (Constraint, Constraint) = (Constraint.Ratio(1, 4), Constraint.Ratio(1, 4))
) extends Widget {

  /** Compute the internal layout of the chart given the area. If the area is too small some elements may be automatically hidden
    */
  def layout(area: Rect): ChartWidget.ChartLayout = {
    var layout = ChartWidget.ChartLayout.default
    if (area.height == 0 || area.width == 0) {
      return layout
    }
    var x = area.left
    var y = area.bottom - 1

    if (xAxis.labels.isDefined && y > area.top) {
      layout = layout.copy(labelX = Some(y))
      y -= 1
    }

    layout = layout.copy(labelY = this.yAxis.labels.map(_ => x))
    x += this.maxWidthOfLabelsLeftOfYAxis(area, this.yAxis.labels.isDefined)

    if (this.xAxis.labels.isDefined && y > area.top) {
      layout = layout.copy(axisX = Some(y))
      y -= 1
    }

    if (this.yAxis.labels.isDefined && x + 1 < area.right) {
      layout = layout.copy(axisY = Some(x))
      x += 1
    }

    if (x < area.right && y > 1) {
      layout = layout.copy(graphArea = Rect(x, area.top, area.right - x, y - area.top + 1))
    }

    this.xAxis.title match {
      case None => ()
      case Some(title) =>
        val w = title.width
        if (w < layout.graphArea.width && layout.graphArea.height > 2) {
          layout = layout.copy(titleX = Some((x + layout.graphArea.width - w, y)))
        }
    }

    this.yAxis.title match {
      case None => ()
      case Some(title) =>
        val w = title.width
        if (w + 1 < layout.graphArea.width && layout.graphArea.height > 2) {
          layout = layout.copy(titleY = Some((x, area.top)))
        }
    }

    this.datasets.map(x => Grapheme(x.name).width).maxOption match {
      case None => ()
      case Some(inner_width) =>
        val legend_width = inner_width + 2
        val legend_height = this.datasets.length + 2
        val max_legend_width = this.hiddenLegendConstraints._1(layout.graphArea.width)
        val max_legend_height = this.hiddenLegendConstraints._2(layout.graphArea.height)
        if (inner_width > 0 && legend_width < max_legend_width && legend_height < max_legend_height) {
          val rect = Rect(
            x = layout.graphArea.right - legend_width,
            y = layout.graphArea.top,
            width = legend_width,
            height = legend_height
          )
          layout = layout.copy(legendArea = Some(rect))
        }
    }
    layout
  }

  def maxWidthOfLabelsLeftOfYAxis(area: Rect, has_y_axis: Boolean): Int = {
    var max_width = this.yAxis.labels.flatMap(labels => labels.map(_.width).maxOption).getOrElse(0)

    this.xAxis.labels.flatMap(labels => labels.headOption) match {
      case None => ()
      case Some(first_x_label) =>
        val first_label_width = Grapheme(first_x_label.content).width
        val width_left_of_y_axis = this.xAxis.labelsAlignment match {
          case Alignment.Left =>
            // The last character of the label should be below the Y-Axis when it exists, not on its left
            val y_axis_offset = if (has_y_axis) 1 else 0
            first_label_width.saturating_sub_unsigned(y_axis_offset)
          case Alignment.Center => first_label_width / 2
          case Alignment.Right  => 0
        }
        max_width = math.max(max_width, width_left_of_y_axis)
    }
    // labels of y axis and first label of x axis can take at most 1/3rd of the total width
    max_width.min(area.width / 3)
  }

  def renderXLabels(buf: Buffer, layout: ChartWidget.ChartLayout, chartArea: Rect, graphArea: Rect): Unit = {
    val y = layout.labelX match {
      case Some(y) => y
      case None    => return
    }
    val labels = this.xAxis.labels.getOrElse(Array.empty[Span])
    val labels_len = labels.length
    if (labels_len < 2) {
      return
    }

    val width_between_ticks = graphArea.width / labels_len

    val label_area = this.firstXLabelArea(y, labels.head.width, width_between_ticks, chartArea, graphArea)

    val label_alignment = this.xAxis.labelsAlignment match {
      case Alignment.Left   => Alignment.Right
      case Alignment.Center => Alignment.Center
      case Alignment.Right  => Alignment.Left
    }

    renderLabel(buf, labels.head, label_area, label_alignment)

    ranges.range(0, labels.length /* first and last not rendered  here */ - 2) { i =>
      val label = labels(i + 1)
      // We add 1 to x (and width-1 below) to leave at least one space before each intermediate labels
      val x = graphArea.left + (i + 1) * width_between_ticks + 1
      val label_area = Rect(x, y, width_between_ticks.saturating_sub_unsigned(1), 1)

      renderLabel(buf, label, label_area, Alignment.Center)
      ()
    }

    val x = graphArea.right - width_between_ticks
    val label_area1 = Rect(x, y, width_between_ticks, 1)
    // The last label should be aligned Right to be at the edge of the graph area
    renderLabel(buf, labels.last, label_area1, Alignment.Right)
    ()
  }

  def firstXLabelArea(y: Int, labelWidth: Int, maxWidthAfterYAxis: Int, chartArea: Rect, graphArea: Rect): Rect = {
    val (min_x, max_x) = this.xAxis.labelsAlignment match {
      case Alignment.Left => (chartArea.left, graphArea.left)
      case Alignment.Center =>
        (
          chartArea.left,
          graphArea.left + maxWidthAfterYAxis.min(labelWidth)
        )
      case Alignment.Right =>
        (
          graphArea.left.saturating_sub_unsigned(1),
          graphArea.left + maxWidthAfterYAxis
        )
    }

    Rect(min_x, y, max_x - min_x, 1)
  }

  def renderLabel(buf: Buffer, label: Span, labelArea: Rect, alignment: Alignment): (Int, Int) = {
    val label_width = label.width
    val bounded_label_width = labelArea.width.min(label_width)

    val x = alignment match {
      case Alignment.Left   => labelArea.left
      case Alignment.Center => labelArea.left + labelArea.width / 2 - bounded_label_width / 2
      case Alignment.Right  => labelArea.right - bounded_label_width
    }

    buf.setSpan(x, labelArea.top, label, bounded_label_width)
  }

  def renderYLabels(buf: Buffer, layout: ChartWidget.ChartLayout, chartArea: Rect, graphArea: Rect): Unit = {
    val x = layout.labelY match {
      case Some(x) => x
      case None    => return
    }
    val labels = this.yAxis.labels.getOrElse(Array.empty[Span])
    val labels_len = labels.length
    ranges.range(0, labels.length) { i =>
      val label = labels(i)
      val dy = i * (graphArea.height - 1) / (labels_len - 1)
      if (dy < graphArea.bottom) {
        val label_area = Rect(
          x,
          graphArea.bottom.saturating_sub_unsigned(1) - dy,
          (graphArea.left - chartArea.left).saturating_sub_unsigned(1),
          1
        )
        renderLabel(buf, label, label_area, this.yAxis.labelsAlignment)
        ()
      }
    }
  }

  override def render(area: Rect, buf: Buffer): Unit = {
    if (area.area == 0) {
      return
    }
    buf.setStyle(area, this.style)
    // Sample the style of the entire widget. This sample will be used to reset the style of
    // the cells that are part of the components put on top of the grah area (i.e legend and
    // axis names).
    val original_style = buf.get(area.left, area.top).style

    val layout = this.layout(area)
    val graph_area = layout.graphArea
    if (graph_area.width < 1 || graph_area.height < 1) {
      return
    }

    this.renderXLabels(buf, layout, area, graph_area)
    this.renderYLabels(buf, layout, area, graph_area)

    layout.axisX match {
      case None => ()
      case Some(y) =>
        ranges.range(graph_area.left, graph_area.right) { x =>
          buf.get(x, y).setSymbol(symbols.line.HORIZONTAL).setStyle(this.xAxis.style)
          ()
        }
    }

    layout.axisY match {
      case None => ()
      case Some(x) =>
        ranges.range(graph_area.top, graph_area.bottom) { y =>
          buf.get(x, y).setSymbol(symbols.line.VERTICAL).setStyle(this.yAxis.style)
          ()
        }
    }

    layout.axisX match {
      case None => ()
      case Some(y) =>
        layout.axisY match {
          case None    => ()
          case Some(x) => buf.get(x, y).setSymbol(symbols.line.BOTTOM_LEFT).setStyle(this.xAxis.style)
        }
    }

    this.datasets.foreach { dataset =>
      CanvasWidget(
        backgroundColor = original_style.bg.getOrElse(Color.Reset),
        xBounds = this.xAxis.bounds,
        yBounds = this.yAxis.bounds,
        marker = dataset.marker
      ) { ctx =>
        val points = Points(coords = dataset.data, color = dataset.style.fg.getOrElse(Color.Reset))
        ctx.draw(points)
        dataset.graphType match {
          case ChartWidget.GraphType.Scatter => ()
          case ChartWidget.GraphType.Line =>
            dataset.data.sliding(2).foreach {
              case Array(one, two) =>
                val line = Line(x1 = one.x, y1 = one.y, x2 = two.x, y2 = two.y, color = dataset.style.fg.getOrElse(Color.Reset))
                ctx.draw(line)
              case _ => ()
            }
        }
      }
        .render(graph_area, buf)
    }

    layout.legendArea match {
      case None => ()
      case Some(legend_area) =>
        buf.setStyle(legend_area, original_style)
        BlockWidget.noChildren(borders = Borders.ALL).render(legend_area, buf)
        ranges.range(0, this.datasets.length) { i =>
          val dataset = this.datasets(i)
          buf.setString(
            legend_area.x + 1,
            legend_area.y + 1 + i,
            dataset.name,
            dataset.style
          )
          ()
        }
    }

    layout.titleX match {
      case None => ()
      case Some((x, y)) =>
        val title = this.xAxis.title.get
        val width = graph_area.right.saturating_sub_unsigned(x)
        buf.setStyle(Rect(x, y, width, height = 1), original_style)
        buf.setSpans(x, y, title, width)
    }

    layout.titleY match {
      case None => ()
      case Some((x, y)) =>
        val title = this.yAxis.title.get
        val width = graph_area.right.saturating_sub_unsigned(x)
        buf.setStyle(Rect(x, y, width, height = 1), original_style)
        buf.setSpans(x, y, title, width)
        ()
    }
  }
}

object ChartWidget {

  /** An X or Y axis for the chart widget
    * @param title
    *   Title displayed next to axis end
    * @param bounds
    *   Bounds for the axis (all data points outside these limits will not be represented)
    * @param labels
    *   A list of labels to put to the left or below the axis
    * @param style
    *   The style used to draw the axis itself
    * @param labelsAlignment
    *   Defines the alignment of the labels of the axis. The alignment behaves differently based on the axis:
    *   - Y-Axis: The labels are aligned within the area on the left of the axis
    *   - X-Axis: The first X-axis label is aligned relative to the Y-axis
    */
  case class Axis(
      title: Option[Spans] = None,
      bounds: Point = Point.Zero,
      labels: Option[Array[Span]] = None,
      style: Style = Style.DEFAULT,
      labelsAlignment: Alignment = Alignment.Left
  )

  object Axis {
    val default: Axis = Axis()
  }

  /** Used to determine which style of graphing to use
    */
  sealed trait GraphType

  object GraphType {

    /** Draw each point
      */
    case object Scatter extends GraphType

    /** Draw each point and lines between each point using the same marker
      */
    case object Line extends GraphType
  }

  /** A group of data points
    *
    * @param name
    *   Name of the dataset (used in the legend if shown)
    * @param data
    *   A reference to the actual data
    * @param marker
    *   Symbol used for each points of this dataset
    * @param graphType
    *   Determines graph type used for drawing points
    * @param style
    *   Style used to plot this dataset
    */
  case class Dataset(
      name: String = "",
      data: Array[Point] = Array.empty,
      marker: symbols.Marker = symbols.Marker.Dot,
      graphType: GraphType = GraphType.Scatter,
      style: Style = Style.DEFAULT
  )

  /** A container that holds all the infos about where to display each elements of the chart (axis, labels, legend, ...).
    *
    * @param titleX
    *   Location of the title of the x axis
    * @param titleY
    *   Location of the title of the y axis
    * @param labelX
    *   Location of the first label of the x axis
    * @param labelY
    *   Location of the first label of the y axis
    * @param axisX
    *   Y coordinate of the horizontal axis
    * @param axisY
    *   X coordinate of the vertical axis
    * @param legendArea
    *   Area of the legend
    * @param graphArea
    *   Area of the graph
    */
  case class ChartLayout(
      titleX: Option[(Int, Int)] = None,
      titleY: Option[(Int, Int)] = None,
      labelX: Option[Int] = None,
      labelY: Option[Int] = None,
      axisX: Option[Int] = None,
      axisY: Option[Int] = None,
      legendArea: Option[Rect] = None,
      graphArea: Rect = Rect.default
  )

  object ChartLayout {
    val default: ChartLayout = ChartLayout()
  }
}

File: tui/src/java/tui/widgets/ChartWidget.java
package tui.widgets;

import java.util.Optional;
import tui.Alignment;
import tui.Borders;
import tui.Buffer;
import tui.Color;
import tui.Constraint;
import tui.Grapheme;
import tui.Point;
import tui.Position;
import tui.Rect;
import tui.Span;
import tui.Spans;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;
import tui.internal.Saturating;
import tui.widgets.canvas.CanvasWidget;
import tui.widgets.canvas.Line;
import tui.widgets.canvas.Points;

/// A widget to plot one or more dataset in a cartesian coordinate system
public final class ChartWidget implements Widget {
  public final Dataset[] datasets;
  public final Optional<BlockWidget> block;
  public final Axis xAxis;
  public final Axis yAxis;
  public final Style style;
  public final HiddenLegendConstraints hiddenLegendConstraints;

  public ChartWidget(
      Dataset[] datasets,
      Optional<BlockWidget> block,
      Axis xAxis,
      Axis yAxis,
      Style style,
      HiddenLegendConstraints hiddenLegendConstraints) {
    this.datasets = datasets;
    this.block = block;
    this.xAxis = xAxis;
    this.yAxis = yAxis;
    this.style = style;
    this.hiddenLegendConstraints = hiddenLegendConstraints;
  }

  public static ChartWidget empty(Dataset[] datasets) {
    return new ChartWidget(
        datasets,
        Optional.empty(),
        Axis.empty(),
        Axis.empty(),
        Style.DEFAULT,
        new HiddenLegendConstraints(new Constraint.Ratio(1, 4), new Constraint.Ratio(1, 4)));
  }

  public ChartWidget withBlock(BlockWidget b) {
    return new ChartWidget(datasets, Optional.of(b), xAxis, yAxis, style, hiddenLegendConstraints);
  }

  public ChartWidget withXAxis(Axis a) {
    return new ChartWidget(datasets, block, a, yAxis, style, hiddenLegendConstraints);
  }

  public ChartWidget withYAxis(Axis a) {
    return new ChartWidget(datasets, block, xAxis, a, style, hiddenLegendConstraints);
  }

  public ChartWidget withStyle(Style s) {
    return new ChartWidget(datasets, block, xAxis, yAxis, s, hiddenLegendConstraints);
  }

  public ChartWidget withHiddenLegendConstraints(HiddenLegendConstraints h) {
    return new ChartWidget(datasets, block, xAxis, yAxis, style, h);
  }

  /// Compute the internal layout of the chart given the area. If the area is too small some elements may be automatically hidden
  public ChartLayout layout(Rect area) {
    ChartLayout layout = ChartLayout.empty();
    if (area.height() == 0 || area.width() == 0) {
      return layout;
    }
    int x = area.left();
    int y = area.bottom() - 1;

    if (xAxis.labels.isPresent() && y > area.top()) {
      layout = layout.withLabelX(Optional.of(y));
      y -= 1;
    }

    layout = layout.withLabelY(yAxis.labels.isPresent() ? Optional.of(x) : Optional.empty());
    x += maxWidthOfLabelsLeftOfYAxis(area, yAxis.labels.isPresent());

    if (xAxis.labels.isPresent() && y > area.top()) {
      layout = layout.withAxisX(Optional.of(y));
      y -= 1;
    }

    if (yAxis.labels.isPresent() && x + 1 < area.right()) {
      layout = layout.withAxisY(Optional.of(x));
      x += 1;
    }

    if (x < area.right() && y > 1) {
      layout = layout.withGraphArea(new Rect(x, area.top(), area.right() - x, y - area.top() + 1));
    }

    if (xAxis.title.isPresent()) {
      Spans title = xAxis.title.get();
      int w = title.width();
      if (w < layout.graphArea.width() && layout.graphArea.height() > 2) {
        layout = layout.withTitleX(Optional.of(new Position(x + layout.graphArea.width() - w, y)));
      }
    }

    if (yAxis.title.isPresent()) {
      Spans title = yAxis.title.get();
      int w = title.width();
      if (w + 1 < layout.graphArea.width() && layout.graphArea.height() > 2) {
        layout = layout.withTitleY(Optional.of(new Position(x, area.top())));
      }
    }

    int innerWidth = -1;
    for (Dataset d : datasets) {
      int w = new Grapheme(d.name).width();
      if (w > innerWidth) innerWidth = w;
    }
    if (innerWidth >= 0) {
      int legendWidth = innerWidth + 2;
      int legendHeight = datasets.length + 2;
      int maxLegendWidth = hiddenLegendConstraints.first.apply(layout.graphArea.width());
      int maxLegendHeight = hiddenLegendConstraints.second.apply(layout.graphArea.height());
      if (innerWidth > 0 && legendWidth < maxLegendWidth && legendHeight < maxLegendHeight) {
        Rect rect =
            new Rect(
                layout.graphArea.right() - legendWidth,
                layout.graphArea.top(),
                legendWidth,
                legendHeight);
        layout = layout.withLegendArea(Optional.of(rect));
      }
    }
    return layout;
  }

  public int maxWidthOfLabelsLeftOfYAxis(Rect area, boolean hasYAxis) {
    int maxWidth = 0;
    if (yAxis.labels.isPresent()) {
      for (Span s : yAxis.labels.get()) {
        int w = s.width();
        if (w > maxWidth) maxWidth = w;
      }
    }

    if (xAxis.labels.isPresent() && xAxis.labels.get().length > 0) {
      Span firstXLabel = xAxis.labels.get()[0];
      int firstLabelWidth = new Grapheme(firstXLabel.content()).width();
      int widthLeftOfYAxis =
          switch (xAxis.labelsAlignment) {
            case Left -> {
              int yAxisOffset = hasYAxis ? 1 : 0;
              yield Saturating.saturatingSubUnsigned(firstLabelWidth, yAxisOffset);
            }
            case Center -> firstLabelWidth / 2;
            case Right -> 0;
          };
      maxWidth = Math.max(maxWidth, widthLeftOfYAxis);
    }
    return Math.min(maxWidth, area.width() / 3);
  }

  public void renderXLabels(Buffer buf, ChartLayout layout, Rect chartArea, Rect graphArea) {
    if (layout.labelX.isEmpty()) return;
    int y = layout.labelX.get();
    Span[] labels = xAxis.labels.orElse(new Span[0]);
    int labelsLen = labels.length;
    if (labelsLen < 2) return;

    int widthBetweenTicks = graphArea.width() / labelsLen;

    Rect labelArea =
        firstXLabelArea(y, labels[0].width(), widthBetweenTicks, chartArea, graphArea);

    Alignment labelAlignment =
        switch (xAxis.labelsAlignment) {
          case Left -> Alignment.Right;
          case Center -> Alignment.Center;
          case Right -> Alignment.Left;
        };

    renderLabel(buf, labels[0], labelArea, labelAlignment);

    Ranges.range(
        0,
        labels.length - 2,
        i -> {
          Span label = labels[i + 1];
          int xx = graphArea.left() + (i + 1) * widthBetweenTicks + 1;
          Rect la = new Rect(xx, y, Saturating.saturatingSubUnsigned(widthBetweenTicks, 1), 1);
          renderLabel(buf, label, la, Alignment.Center);
        });

    int xx = graphArea.right() - widthBetweenTicks;
    Rect labelArea1 = new Rect(xx, y, widthBetweenTicks, 1);
    renderLabel(buf, labels[labels.length - 1], labelArea1, Alignment.Right);
  }

  public Rect firstXLabelArea(
      int y, int labelWidth, int maxWidthAfterYAxis, Rect chartArea, Rect graphArea) {
    int minX;
    int maxX;
    switch (xAxis.labelsAlignment) {
      case Left -> {
        minX = chartArea.left();
        maxX = graphArea.left();
      }
      case Center -> {
        minX = chartArea.left();
        maxX = graphArea.left() + Math.min(maxWidthAfterYAxis, labelWidth);
      }
      case Right -> {
        minX = Saturating.saturatingSubUnsigned(graphArea.left(), 1);
        maxX = graphArea.left() + maxWidthAfterYAxis;
      }
      default -> {
        minX = 0;
        maxX = 0;
      }
    }
    return new Rect(minX, y, maxX - minX, 1);
  }

  public Position renderLabel(Buffer buf, Span label, Rect labelArea, Alignment alignment) {
    int labelWidth = label.width();
    int boundedLabelWidth = Math.min(labelArea.width(), labelWidth);

    int x =
        switch (alignment) {
          case Left -> labelArea.left();
          case Center -> labelArea.left() + labelArea.width() / 2 - boundedLabelWidth / 2;
          case Right -> labelArea.right() - boundedLabelWidth;
        };

    return buf.setSpan(x, labelArea.top(), label, boundedLabelWidth);
  }

  public void renderYLabels(Buffer buf, ChartLayout layout, Rect chartArea, Rect graphArea) {
    if (layout.labelY.isEmpty()) return;
    int x = layout.labelY.get();
    Span[] labels = yAxis.labels.orElse(new Span[0]);
    int labelsLen = labels.length;
    Ranges.range(
        0,
        labels.length,
        i -> {
          Span label = labels[i];
          int dy = i * (graphArea.height() - 1) / (labelsLen - 1);
          if (dy < graphArea.bottom()) {
            Rect la =
                new Rect(
                    x,
                    Saturating.saturatingSubUnsigned(graphArea.bottom(), 1) - dy,
                    Saturating.saturatingSubUnsigned(graphArea.left() - chartArea.left(), 1),
                    1);
            renderLabel(buf, label, la, yAxis.labelsAlignment);
          }
        });
  }

  @Override
  public void render(Rect area, Buffer buf) {
    if (area.area() == 0) {
      return;
    }
    buf.setStyle(area, style);
    Style originalStyle = buf.get(area.left(), area.top()).style();

    Rect chartArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      chartArea = innerArea;
    } else {
      chartArea = area;
    }

    ChartLayout layout = layout(chartArea);
    Rect graphArea = layout.graphArea;
    if (graphArea.width() < 1 || graphArea.height() < 1) {
      return;
    }

    renderXLabels(buf, layout, chartArea, graphArea);
    renderYLabels(buf, layout, chartArea, graphArea);

    if (layout.axisX.isPresent()) {
      int y = layout.axisX.get();
      Ranges.range(
          graphArea.left(),
          graphArea.right(),
          x -> buf.get(x, y).setSymbol(Symbols.line.HORIZONTAL).setStyle(xAxis.style));
    }

    if (layout.axisY.isPresent()) {
      int x = layout.axisY.get();
      Ranges.range(
          graphArea.top(),
          graphArea.bottom(),
          y -> buf.get(x, y).setSymbol(Symbols.line.VERTICAL).setStyle(yAxis.style));
    }

    if (layout.axisX.isPresent() && layout.axisY.isPresent()) {
      int y = layout.axisX.get();
      int x = layout.axisY.get();
      buf.get(x, y).setSymbol(Symbols.line.BOTTOM_LEFT).setStyle(xAxis.style);
    }

    for (Dataset dataset : datasets) {
      CanvasWidget cw =
          new CanvasWidget(
                  Optional.empty(),
                  xAxis.bounds,
                  yAxis.bounds,
                  style.bg().orElse(Color.Reset),
                  dataset.marker,
                  ctx -> {
                    Points points =
                        new Points(dataset.data, dataset.style.fg().orElse(Color.Reset));
                    ctx.draw(points);
                    if (dataset.graphType == GraphType.Line) {
                      Point[] data = dataset.data;
                      for (int i = 0; i + 1 < data.length; i++) {
                        Point one = data[i];
                        Point two = data[i + 1];
                        Line line =
                            new Line(
                                one.x(),
                                one.y(),
                                two.x(),
                                two.y(),
                                dataset.style.fg().orElse(Color.Reset));
                        ctx.draw(line);
                      }
                    }
                  });
      cw.render(graphArea, buf);
    }

    if (layout.legendArea.isPresent()) {
      Rect legendArea = layout.legendArea.get();
      buf.setStyle(legendArea, originalStyle);
      BlockWidget.empty().withBorders(Borders.ALL).render(legendArea, buf);
      Ranges.range(
          0,
          datasets.length,
          i -> {
            Dataset dataset = datasets[i];
            buf.setString(
                legendArea.x() + 1, legendArea.y() + 1 + i, dataset.name, dataset.style);
          });
    }

    if (layout.titleX.isPresent()) {
      Position pos = layout.titleX.get();
      int x = pos.x();
      int y = pos.y();
      Spans title = xAxis.title.get();
      // Use the title's own width so the chart's overall style isn't
      // applied to the rest of the row (which would override the dataset
      // style of the top graph line).
      int width = title.width();
      buf.setStyle(new Rect(x, y, width, 1), originalStyle);
      buf.setSpans(x, y, title, width);
    }

    if (layout.titleY.isPresent()) {
      Position pos = layout.titleY.get();
      int x = pos.x();
      int y = pos.y();
      Spans title = yAxis.title.get();
      int width = title.width();
      buf.setStyle(new Rect(x, y, width, 1), originalStyle);
      buf.setSpans(x, y, title, width);
    }
  }

  public record HiddenLegendConstraints(Constraint first, Constraint second) {}

  /// An X or Y axis for the chart widget
  public static final class Axis {
    public final Optional<Spans> title;
    public final Point bounds;
    public final Optional<Span[]> labels;
    public final Style style;
    public final Alignment labelsAlignment;

    public Axis(
        Optional<Spans> title,
        Point bounds,
        Optional<Span[]> labels,
        Style style,
        Alignment labelsAlignment) {
      this.title = title;
      this.bounds = bounds;
      this.labels = labels;
      this.style = style;
      this.labelsAlignment = labelsAlignment;
    }

    public static Axis empty() {
      return new Axis(
          Optional.empty(), Point.Zero, Optional.empty(), Style.DEFAULT, Alignment.Left);
    }

    public Axis withTitle(Spans t) {
      return new Axis(Optional.of(t), bounds, labels, style, labelsAlignment);
    }

    public Axis withBounds(Point b) {
      return new Axis(title, b, labels, style, labelsAlignment);
    }

    public Axis withLabels(Span[] l) {
      return new Axis(title, bounds, Optional.of(l), style, labelsAlignment);
    }

    public Axis withStyle(Style s) {
      return new Axis(title, bounds, labels, s, labelsAlignment);
    }

    public Axis withLabelsAlignment(Alignment a) {
      return new Axis(title, bounds, labels, style, a);
    }
  }

  /// Used to determine which style of graphing to use
  public enum GraphType {
    /// Draw each point
    Scatter,
    /// Draw each point and lines between each point using the same marker
    Line
  }

  /// A group of data points
  public static final class Dataset {
    public final String name;
    public final Point[] data;
    public final Symbols.Marker marker;
    public final GraphType graphType;
    public final Style style;

    public Dataset(
        String name,
        Point[] data,
        Symbols.Marker marker,
        GraphType graphType,
        Style style) {
      this.name = name;
      this.data = data;
      this.marker = marker;
      this.graphType = graphType;
      this.style = style;
    }

    public static Dataset empty() {
      return new Dataset("", new Point[0], Symbols.Marker.Dot, GraphType.Scatter, Style.DEFAULT);
    }

    public Dataset withName(String n) {
      return new Dataset(n, data, marker, graphType, style);
    }

    public Dataset withData(Point[] d) {
      return new Dataset(name, d, marker, graphType, style);
    }

    public Dataset withMarker(Symbols.Marker m) {
      return new Dataset(name, data, m, graphType, style);
    }

    public Dataset withGraphType(GraphType g) {
      return new Dataset(name, data, marker, g, style);
    }

    public Dataset withStyle(Style s) {
      return new Dataset(name, data, marker, graphType, s);
    }
  }

  /// A container that holds all the infos about where to display each elements of the chart.
  public static final class ChartLayout {
    public final Optional<Position> titleX;
    public final Optional<Position> titleY;
    public final Optional<Integer> labelX;
    public final Optional<Integer> labelY;
    public final Optional<Integer> axisX;
    public final Optional<Integer> axisY;
    public final Optional<Rect> legendArea;
    public final Rect graphArea;

    public ChartLayout(
        Optional<Position> titleX,
        Optional<Position> titleY,
        Optional<Integer> labelX,
        Optional<Integer> labelY,
        Optional<Integer> axisX,
        Optional<Integer> axisY,
        Optional<Rect> legendArea,
        Rect graphArea) {
      this.titleX = titleX;
      this.titleY = titleY;
      this.labelX = labelX;
      this.labelY = labelY;
      this.axisX = axisX;
      this.axisY = axisY;
      this.legendArea = legendArea;
      this.graphArea = graphArea;
    }

    public static ChartLayout empty() {
      return new ChartLayout(
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Rect.DEFAULT);
    }

    public ChartLayout withTitleX(Optional<Position> v) {
      return new ChartLayout(v, titleY, labelX, labelY, axisX, axisY, legendArea, graphArea);
    }

    public ChartLayout withTitleY(Optional<Position> v) {
      return new ChartLayout(titleX, v, labelX, labelY, axisX, axisY, legendArea, graphArea);
    }

    public ChartLayout withLabelX(Optional<Integer> v) {
      return new ChartLayout(titleX, titleY, v, labelY, axisX, axisY, legendArea, graphArea);
    }

    public ChartLayout withLabelY(Optional<Integer> v) {
      return new ChartLayout(titleX, titleY, labelX, v, axisX, axisY, legendArea, graphArea);
    }

    public ChartLayout withAxisX(Optional<Integer> v) {
      return new ChartLayout(titleX, titleY, labelX, labelY, v, axisY, legendArea, graphArea);
    }

    public ChartLayout withAxisY(Optional<Integer> v) {
      return new ChartLayout(titleX, titleY, labelX, labelY, axisX, v, legendArea, graphArea);
    }

    public ChartLayout withLegendArea(Optional<Rect> v) {
      return new ChartLayout(titleX, titleY, labelX, labelY, axisX, axisY, v, graphArea);
    }

    public ChartLayout withGraphArea(Rect v) {
      return new ChartLayout(titleX, titleY, labelX, labelY, axisX, axisY, legendArea, v);
    }
  }
}
